oxidation/libparsec/crates/crypto/src/common/sequester.rs

```diff
@@ -0,0 +1,57 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 (eventually AGPL-3.0) 2016-present Scille SAS
+
+use crate::{CryptoError, CryptoResult};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[repr(usize)]
+pub enum SequesterKeySize {
+    _1024Bits = 1024,
+    _2048Bits = 2048,
+    _3072Bits = 3072,
+    _4096Bits = 4096,
+}
+
+pub(crate) trait EnforceSerialize {
+    const ALGORITHM: &'static [u8];
+    fn size_in_bytes(&self) -> usize;
+
+    /// Here we avoid uncessary allocation & enforce output has `key_size`
+    fn serialize(&self, output: &[u8], data: &[u8]) -> Vec<u8> {
+        assert!(output.len() <= self.size_in_bytes());
```

Comment:
tl;dr: you don't need trait, you (basically) never need trait. If you think you need a trait go on Slack and explain why you need one.

I find this an interesting case why Trait add complexity for no benefit most of the time (basically anytime you don't need a common interface on multiple struct at runtime)

- In the Python code we had together the generation of `output` and it use to produce the final serialized payload
- We had to replace this code in Rust by 2 implementations (RustCrypto + Sodiumoxide)
- 2 implementation ? then we want to factorize the code, and hence we separate the generation of `output` (which is specific to RustCrypto/Sodiumoxide) from the final serialization
- To do this factorization, we introduce a trait. Now we have a method that only deal with the final serialization and is only needed in a very specific place.

On the other let me introduce you to the all mighty... function✨ 😮 😍  

```rust
// The function is not available outside of the crypto crate: no risk of misuse !
pub(crate) fn sequester_armor_output(output: &[u8], data: &[u8], key_size_in_bytes: usize, algo: &str) -> vec<u8> {
    assert!(output.len() <= key_size_in_bytes);
    // Do the serialization stuff
    res
}
impl SequesterSigningKeyDer {
  pub fn sign(data: &[u8]) -> vec<u8> {
    let output = do_actual_signature(data);
    // The code is more readable given we don't have to jump to the implementation of the trait !
    sequester_armor_output(output, data, self.size_in_bytes(), "RSAES-OAEP-XSALSA20-POLY1305")
  }
}
```

For more advanced case (typically when different call sites uses different types), a macro could be used instead of the function \o/

@FirelightFlagboy @vxgmichel what do you think ?

Reply:
@FirelightFlagboy Or ask there, if we should keep this comment